src/library/features/crates/cratefeaturehelper.cpp

```diff
@@ -67,9 +71,71 @@ CrateId CrateFeatureHelper::createEmptyCrate() {
     }
 
     CrateId newCrateId;
-    if (m_pTrackCollection->insertCrate(newCrate, &newCrateId)) {
+    if (m_pCrates->insertCrate(newCrate, &newCrateId)) {
         DEBUG_ASSERT(newCrateId.isValid());
         newCrate.setId(newCrateId);
+        m_pCrates->hierarchy().initClosureForCrate(newCrateId);
+        m_pCrates->hierarchy().generateCratePaths(newCrate);
+        qDebug() << "Created new crate" << newCrate;
+    } else {
+        DEBUG_ASSERT(!newCrateId.isValid());
+        qWarning() << "Failed to create new crate"
+                << "->"  << newCrate.getName();
+        QMessageBox::warning(
+                nullptr,
+                tr("Creating Crate Failed"),
+                tr("An unknown error occurred while creating crate: ") + newCrate.getName());
+    }
+    return newCrateId;
+}
+
+CrateId CrateFeatureHelper::createEmptySubrate(const Crate& parent) {
+    const QString proposedCrateName =
+            proposeNameForNewCrate(tr("New Subcrate"));
+    Crate newCrate;
+    while (!newCrate.hasName()) {
+        bool ok = false;
+        newCrate.parseName(
+                QInputDialog::getText(
+                        nullptr,
+                        tr("Create New Crate"),
+                        tr("Enter name for new crate:"),
+                        QLineEdit::Normal,
+                        proposedCrateName,
+                        &ok));
+        if (!ok) {
+            return CrateId();
+        }
+        if (!newCrate.hasName()) {
+            QMessageBox::warning(
+                    nullptr,
+                    tr("Creating Crate Failed"),
+                    tr("A crate cannot have a blank name."));
+            continue;
+        }
+
+        // select name from crateClosure join crates on id = childId group by childId having count(*) = 1;
+        // check if it has parent, if not compare with names from above
+        // in closure.
+        // Else split path into tokens and compare to those
+        if (m_pCrates->hierarchy().tokenizeCratePath(parent.getId()).contains(newCrate.getName())) {
```

Comment:
This is very similar to the logic above for root level crates. Could both cases be handled by a new public method of CrateHierarchy?